lib/Service/SocialApiService.php

```diff
@@ -0,0 +1,224 @@
+				}
+			}
+		}
+		if (is_null($type)) {
+			return null;
+		}
+		return $type;
+	}
+
+
+	/**
+	 * @NoAdminRequired
+	 *
+	 * Adds/updates photo for contact
+	 *
+	 * @param {pointer} contact reference to the contact to update
+	 * @param {string} imageType the image type of the photo
+	 * @param {string} photo the photo as base64 string
+	 */
+	protected function addPhoto(array &$contact, string $imageType, string $photo) {
+
+		$version = $contact['VERSION'];
+
+		if (!empty($contact['PHOTO'])) {
+			// overwriting without notice!
+		}
+
+		if ($version >= 4.0) {
+			// overwrite photo
+			$contact['PHOTO'] = "data:" . $imageType . ";base64," . $photo;
+		}
+
+		elseif ($version >= 3.0) {
+			// add new photo
+			$imageType = str_replace('image/', '', $imageType);
+			$contact['PHOTO;ENCODING=b;TYPE=' . $imageType . ';VALUE=BINARY'] = $photo;
+
+			// remove previous photo (necessary as new attribute is not equal to 'PHOTO')
+			$contact['PHOTO'] = '';
+		}
+	}
+
+
+	/**
+	 * @NoAdminRequired
+	 *
+	 * Gets the addressbook of an addressbookId
+	 *
+	 * @param {String} addressbookId the identifier of the addressbook
+	 *
+	 * @returns {IAddressBook} the corresponding addressbook or null
+	 */
+	protected function getAddressBook(string $addressbookId) : ?IAddressBook {
+		$addressBook = null;
+		$addressBooks = $this->manager->getUserAddressBooks();
+		foreach($addressBooks as $ab) {
+			if ($ab->getUri() === $addressbookId) {
+				$addressBook = $ab;
+			}
+		}
+		return $addressBook;
+	}
+
+
+	/**
+	 * @NoAdminRequired
+	 *
+	 * Retrieves social profile data for a contact and updates the entry
+	 *
+	 * @param {String} addressbookId the addressbook identifier
+	 * @param {String} contactId the contact identifier
+	 * @param {String} network the social network to use (if unkown: take first match)
+	 *
+	 * @returns {JSONResponse} an empty JSONResponse with respective http status code
+	 */
+	public function updateContact(string $addressbookId, string $contactId, string $network) : JSONResponse {
+
+		$url = null;
+
+		try {
+			// get corresponding addressbook
+			$addressBook = $this->getAddressBook($addressbookId);
+			if (is_null($addressBook)) {
+				return new JSONResponse([], Http::STATUS_BAD_REQUEST);
+			}
+
+			// search contact in that addressbook, get social data
+			$contact = $addressBook->search($contactId, ['UID'], ['types' => true])[0];
+			if (!isset($contact['X-SOCIALPROFILE'])) {
+				return new JSONResponse([], Http::STATUS_PRECONDITION_FAILED);
+			}
+			$socialprofiles = $contact['X-SOCIALPROFILE'];
+			// retrieve data
+			$url = $this->socialProvider->getSocialConnector($socialprofiles, $network);
+
+			if (empty($url)) {
+				return new JSONResponse([], Http::STATUS_BAD_REQUEST);
+			}
+
+			$opts = [
+				"http" => [
+					"method" => "GET",
+					"header" => "User-Agent: Nextcloud Contacts App"
+				]
+			];
+			$context = stream_context_create($opts);
+			$socialdata = file_get_contents($url, false, $context);
```

Comment:
:warning: I would not download a file with this but use Nextcloud's HTTP service to get a guzzle HTTP client that also knows how to use proxies and so on.

Reply:
Thank you for the good hint, I was not aware of that functionality!